04-onion-routing.md

```diff
@@ -463,6 +463,163 @@ Having the shared secrets of all intermediate nodes it can unwrap the packet unt
 
 The association between forward and return packet is handled outside of the protocol, e.g., by association to an HTLC in a payment channel.
 
+### Failure Codes
+
+The following failure codes are supported.  A node MUST select one of
+these failure codes when creating an error message, and MUST include the
+appropriate `additional` data.
+
+In the case of more than one error, a node SHOULD select the first one
+listed.
+
+The top byte of the failure code can be read as a set of flags:
+* 0x80000000 (BADONION): unparsable onion, encrypted by previous node.
+* 0x80000000 (PERM): permanent failure (otherwise transient)
+* 0x40000000 (NODE): node failure (otherwise channel)
+* 0x20000000 (UPDATE): new channel update enclosed
+
+Any node may return one of the following errors:
+
+If the realm byte is unknown:
+
+1. type: PERM|1 (`invalid_realm`)
+
+If an otherwise unspecified transient error occurs for the entire
+node:
+
+1. type: NODE|2 (`temporary_node_failure`)
+
+If an otherwise unspecified permanent error occurs for the entire
+node:
+
+1. type: PERM|NODE|2 (`permanent_node_failure`)
+
+If a node has requirement advertised in its `node_announcement`
+`features` which were not present in the onion:
+
+1. type: PERM|NODE|3 (`required_node_feature_missing`)
+
+A forwarding node MAY return one of the following errors, the final
+node MUST NOT:
+
+If the onion version byte is unknown:
+
+1. type: BADONION|PERM|4 (`invalid_onion_version`)
+2. data:
+   * [32:sha256-of-onion]
+
+If the ephemeral key in the onion is invalid:
+
+1. type: BADONION|PERM|5 (`invalid_onion_key`)
+2. data:
+   * [32:sha256-of-onion]
```

Comment:
Yes, this just covers grossly malformed: if onion key doesn't start with 0x2 or 0x3, it's invalid, for example; we can't even pretend to use it.

We should add another BADONION code for bad HMAC.